### PR TITLE
Standalone Search Widget

### PR DESCRIPTION
**Pull request by yih-en**
### Basic Info
Search Widget that allow user to search based on zAPIness endpoints

[Live Demo](http://zapi-widget.herokuapp.com/)

screen shot 2017-04-18 at 3 31 37 pm

___

### Basic Usage

Load Search Widget in any webpage by embedding the following javascripts:

```javascript
<script type="text/javascript">
  var script = document.createElement('script');
  script.src = "https://govwizely.github.io/zAPIness-search-widget/static/js/main.fdfe92ee.js";
  document.getElementsByTagName('head')[0].appendChild(script);

  window.onload = function() {
    //Create div element
    var div = document.createElement("div");
    div.setAttribute("id", "root");
    document.body.appendChild(div);

    // Initiate the search widget with endpoint
    SearchWidget.new({
      endpoint: "[zAPIness endpoint]"
    });
  }
</script>
```
___

### Stacks Used

* React + REDUX
* [Immutable-js](https://facebook.github.io/immutable-js/) - For persistent immutable data
* [Radium](http://formidable.com/open-source/radium/) - For inline css stylings
* [Redux Form](http://redux-form.com/6.6.3/)
* Webpack 
* Babel
* Eslint - based on the latest airbnb config
* Jest - For testing
* [react-script](https://www.npmjs.com/package/react-scripts) - To build production based JS files
___

### Basic Workflow

When Search Widget is initialized, the endpoint will be stored as the main endpoint in ```actions/api.js```. Any requests will be made according to this endpoint after initialization. Redux store will be initialized with the following initial states:

```
const initialState = Map({
  categories: [],
  keyword: '',
  offset: 0,
  pageNum: 1,
  filters: [],
  data: undefined,
  error: []
})
```

* ___categories___: 
	* This field will be populated right before the React App is mounted.
	* It grabs values from data.aggregations passed from the endpoint.
	* It is used in creating filters in the search widget
* ___keyword___: 
	* Store user input.
	* ```GET``` requests will be made when user key in the search input field. The onchange event is debounced to avoid redundant requests to backend server. Refer to ```src/components/Form.js:L44```
* ___offset___:
	* Used in Pagination that determine the data offset according to the selected page number
	* This field is used in building params in ```GET``` request. To access the info in subsequent page, we need to pass in ```?offset=[num]```

* ___pageNum___:
	* Store user input on page number
	* Important in building params 
* ___filters___:
	* Used in query builder
	* Ideally, the structure of search query should be:
	```
    params = {
       "q": "office",
       "offset": 0,
       "filter": {
         agency: ['ausaid', 'australian national audit office (anao)'],
         category: ['architectural engineering']
       }
     }
     ```
     * The type and values of the filters are dictated by user input

* ___data___:
	* Result returned by the endpoint

* ___error___:
	* Store any errors returned by the endpoint

### CODE REVIEW

File: src/components/Drawer.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  render() {
+    const {
+      cells,
+      label
+    } = this.props
+
+    return (
+      <a
+        href="{undefined}"
```
Comment (viclim) [inline src/components/Drawer.js:35]: @yih-en I have never seen this before. Is it different from `javascript:void(0)`?

**Comment by viclim** [comment]
@yih-en 
A search seems to hit the API multiple times. [screenshot](http://recordit.co/h1r5xfI613)

**Comment by yih-en** [comment]
@viclim, it's expected to have 2 API calls in your case. The first time the widget loads, an API call will be made to retrieve all the filters available for the dropdown in advanced search. Then, when you key in the input, another API call will be made once you stop typing/interacting with the search bar.

File: src/components/Drawer.js (hunk 1 above)
Comment (yih-en) [reply src/components/Drawer.js:35]: This is related to the ESLint: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/href-no-hash.md. Also, using ```javascript:void(0)``` violates this rule: http://eslint.org/docs/rules/no-script-url.

**Comment by viclim** [comment]
That makes sense.

File: src/actions/__tests__/range-test.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  it('returns range between 2 and totalPage if totalPage' +
+     'is larger than preset range', () => {
+    expect(getRange(1, 1, 3)).toEqual([2])
+  })
+
+  it('returns range before and after current number according to set limit', () => {
+    expect(getRange(4, 10, 3)).toEqual([3, 4, 5])
+  })
+
+  it('sets initial to 2 if current number is 1', () => {
```
Comment (blackcofla) [inline src/actions/__tests__/range-test.js:13]: Seems like at the end, this test exercise the same flow than first test?

File: src/actions/range.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { floor, range } from 'lodash'
+
+const noPrev = (current, max) => current - max <= 1
+
+const noNext = (current, totalPage, max) => current + max >= totalPage
+
+const getRange = (activePage, totalPage, totalNumButton) => {
```
Comment (blackcofla) [inline src/actions/range.js:7]: What about validating that the 3 arguments are bigger than 0?

File: src/components/Pagination.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                e.preventDefault()
+                onSelect(activePage + 1)
+              }}
+            >
+              Next ›
+            </List>
+          }
+
+          <List
+            className="previous-page"
```
Comment (blackcofla) [inline src/components/Pagination.js]: This style shouldn't be "next-page"?

File: src/components/__tests__/Filter-test.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import React from 'react'
+import { shallow } from 'enzyme'
+
+import { Filter } from '../Filter'
+
+const { Map } = require('immutable')
+
+describe('components/Filter', () => {
```
Comment (blackcofla) [inline src/components/__tests__/Filter-test.js:8]: Shouldn't we test if the handlers are called properly like in previous test?

File: src/components/__tests__/Pagination-test.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        activePage={activePage}
+        onSelect={onSelect}
+      />
+    )
+
+    expect(pagination.containsMatchingElement(
+      <div>
+        <ul>
+          <li><a>« First</a></li>
+          <li><a>« Prev</a></li>
```
Comment (blackcofla) [inline src/components/__tests__/Pagination-test.js:26]: I'm confused, shouldn't the previous be rendered with a less than sign like:

< Prev

**Review by blackcofla — CHANGES_REQUESTED** [review]

File: src/components/__tests__/Filter-test.js (hunk 5 above)
Comment (yih-en) [reply src/components/__tests__/Filter-test.js:8]: Yeah. Let me add that.

File: src/components/Pagination.js (hunk 4 above)
Comment (yih-en) [reply src/components/Pagination.js]: My bad. As we are using inline-style, the className doesn't matter in this case. ```className``` are mainly used in tests.

File: src/actions/range.js (hunk 3 above)
Comment (yih-en) [reply src/actions/range.js:7]: sure

File: src/actions/__tests__/range-test.js (hunk 2 above)
Comment (yih-en) [reply src/actions/__tests__/range-test.js:13]: Not quite the same. The first one test the scenario when the total page returned is less than our preset range, let's say, 3. Then, it should return the range between the current page and last page.

**Pull request merged**